`azurerm_windows[linux]_function[web]_app`, `azurerm_windows[linux]_function[web]_app_slot` - add `description` property for `ip_restriction` block

xiaxyi opened the pull request:
enable user to add description for their ip restriction rules:
```
"ipSecurityRestrictions": [
            {
                "ipAddress": "10.10.10.10/32",
                "action": "Allow",
                "tag": "Default",
                "priority": 123,
                "name": "test-restriction",
                "description": "Allow ip address 10.10.10.10/32",
                "headers": {
                    "x-azure-fdid": [
                        "55ce4ed1-4b06-4bf1-b40e-4638452104da"
                    ],
                    "x-fd-healthprobe": [
                        "1"
                    ],
                    "x-forwarded-for": [
                        "9.9.9.9/32",
                        "2002::1234:abcd:ffff:c0a8:101/64"
                    ],
                    "x-forwarded-host": [
                        "example.com"
                    ]
                }
            },
```

The failed acc tests are not relate to the property added.

katbyte commented:
Now that the API has been updated and resources migrated can you please resolve the conflicts here?

katbyte reviewed (changes requested):
tests are failing:
```
------- Stdout: -------
=== RUN   TestAccLinuxFunctionApp_withConnectionStringsUpdate
=== PAUSE TestAccLinuxFunctionApp_withConnectionStringsUpdate
=== CONT  TestAccLinuxFunctionApp_withConnectionStringsUpdate
    testcase.go:113: Step 1/8 error: Error running apply: exit status 1
        
        Error: creating App Service Plan (Subscription: "*******"
        Resource Group Name: "acctestRG-LFA-240209191514290172"
        Server Farm Name: "acctestASP-240209191514290172"): performing CreateOrUpdate: unexpected status 409 with response: {"Code":"Conflict","Message":"The pricing tier 'Standard' is not allowed in this resource group. Use this link to learn more: https://aka.ms/app-service-plan-scale-not-allowed","Target":null,"Details":[{"Message":"The pricing tier 'Standard' is not allowed in this resource group. Use this link to learn more: https://aka.ms/app-service-plan-scale-not-allowed"},{"Code":"Conflict"},{"ErrorEntity":{"ExtendedCode":"04114","MessageTemplate":"The pricing tier '{0}' is not allowed in this resource group. Use this link to learn more: https://aka.ms/app-service-plan-scale-not-allowed","Parameters":["Standard"],"Code":"Conflict","Message":"The pricing tier 'Standard' is not allowed in this resource group. Use this link to learn more: https://aka.ms/app-service-plan-scale-not-allowed"}}],"Innererror":null}
        
          with azurerm_service_plan.test,
          on terraform_plugin_test.tf line 35, in resource "azurerm_service_plan" "test":
          35: resource "azurerm_service_plan" "test" {
        
        creating App Service Plan (Subscription:
        "*******"
        Resource Group Name: "acctestRG-LFA-240209191514290172"
        Server Farm Name: "acctestASP-240209191514290172"): performing
        CreateOrUpdate: unexpected status 409 with response:
        {"Code":"Conflict","Message":"The pricing tier 'Standard' is not allowed in
        this resource group. Use this link to learn more:
        https://aka.ms/app-service-plan-scale-not-allowed","Target":null,"Details":[{"Message":"The
        pricing tier 'Standard' is not allowed in this resource group. Use this link
        to learn more:
        https://aka.ms/app-service-plan-scale-not-allowed"},{"Code":"Conflict"},{"ErrorEntity":{"ExtendedCode":"04114","MessageTemplate":"The
        pricing tier '{0}' is not allowed in this resource group. Use this link to
        learn more:
        https://aka.ms/app-service-plan-scale-not-allowed","Parameters":["Standard"],"Code":"Conflict","Message":"The
        pricing tier 'Standard' is not allowed in this resource group. Use this link
        to learn more:
        https://aka.ms/app-service-plan-scale-not-allowed"}}],"Innererror":null}
--- FAIL: TestAccLinuxFunctionApp_withConnectionStringsUpdate (71.54s)
FAIL
```

jackofallops commented on internal/services/appservice/helpers/shared_schema.go:
```diff
@@ -1157,6 +1171,12 @@ func ExpandIpRestrictions(restrictions []IpRestriction) (*[]webapps.IPSecurityRe
 		restriction.Priority = pointer.To(v.Priority)
 
+		if v.Description != "" {
+			restriction.Description = utils.String(v.Description)
+		}
+
+		restriction.Priority = utils.Int64(v.Priority)
```
Can we replace these with `pointer.To()` as we're aiming to remove the `utils` package?
```suggestion
			restriction.Description = pointer.To(v.Description)
		}

		restriction.Priority = pointer.To(v.Priority)
```

jackofallops commented on internal/services/appservice/helpers/shared_schema.go:
```diff
@@ -1492,6 +1512,11 @@ func FlattenIpRestrictions(ipRestrictionsList *[]webapps.IPSecurityRestriction)
 			ipRestriction.Action = *v.Action
 		}
 
+		ipRestriction.Headers = flattenIpRestrictionHeaders(pointer.From(v.Headers))
```
This line appears to have been duplicated? (L1520 below)

jackofallops reviewed (changes requested):
Hi @xiaxyi - Thanks for this, just a couple comments to take a look at.  Can you also ensure that at least one test covers removing a description from the IP Restriction to ensure that use case is covered?

Thanks!

xiaxyi commented:
Thanks @jackofallops for the review, updated the test case and removed the duplicated lines. 

Tests are good and the failed one isn't related to the code change in the PR.
![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/92154856/a1222e07-e6a4-42d4-8ea9-36ced8fe3792)

stephybun reviewed (approved):
Thanks @xiaxyi LGTM 💯
